[Bolt Driver 4.0.0] Driver hangs if connection URI is not reachable

This has first been reported [here](https://github.com/liquigraph/liquigraph/issues/375).
Here is a minimal snippet to reproduce the issue (provided the Neo4j JDBC Bolt driver 4.0.0 is in the classpath):

```java
package org.liquigraph.core;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;

public class Test {

    public static void main(String[] args) {
        getConnection("jdbc:neo4j:bolt://example.com", new Properties());
    }

    public static Connection getConnection(String uri, Properties properties) {
        try {
            return DriverManager.getConnection(uri, properties);
        } catch (SQLException e) {
            throw propagate(e);
        }
    }

    public static RuntimeException propagate(Throwable t) {
        if (t instanceof RuntimeException) {
            return (RuntimeException) t;
        }
        return new RuntimeException(t);
    }
}
```

I expect this program to run to completion. It just hangs instead.

I can reproduce the failure but the VM properly terminates (note I had to send a query otherwise the program would not crash):

```java
package org.liquigraph.core;

import org.neo4j.driver.Driver;
import org.neo4j.driver.GraphDatabase;
import org.neo4j.driver.Result;
import org.neo4j.driver.Session;

public class Test2 {

    public static void main(String[] args) {
        try (Driver driver = GraphDatabase.driver("bolt://example.com");
             Session session = driver.session()) {

            Result result = session.run("RETURN n");
            while (result.hasNext()) {
                System.out.println(result.next());
            }
        }
    }
}
```
Here are the logs in that case:
```
INFO: Direct driver instance 1907431275 created for server address example.com:7687
May 04, 2020 2:15:25 PM org.neo4j.driver.internal.logging.JULogger info
INFO: Closing driver instance 1907431275
May 04, 2020 2:15:25 PM org.neo4j.driver.internal.logging.JULogger info
INFO: Closing connection pool towards example.com:7687
Exception in thread "main" org.neo4j.driver.exceptions.ServiceUnavailableException: Unable to connect to example.com:7687, ensure the database is running and that there is a working network connection to it.
	at org.neo4j.driver.internal.util.Futures.blockingGet(Futures.java:143)
	at org.neo4j.driver.internal.InternalSession.run(InternalSession.java:69)
	at org.neo4j.driver.internal.InternalSession.run(InternalSession.java:51)
	at org.neo4j.driver.internal.AbstractQueryRunner.run(AbstractQueryRunner.java:37)
	at org.neo4j.driver.internal.AbstractQueryRunner.run(AbstractQueryRunner.java:55)
	at org.liquigraph.core.Test2.main(Test2.java:14)
	Suppressed: org.neo4j.driver.internal.util.ErrorUtil$InternalExceptionCause
		at org.neo4j.driver.internal.async.connection.ChannelConnectedListener.databaseUnavailableError(ChannelConnectedListener.java:76)
		at org.neo4j.driver.internal.async.connection.ChannelConnectedListener.operationComplete(ChannelConnectedListener.java:70)
		at org.neo4j.driver.internal.async.connection.ChannelConnectedListener.operationComplete(ChannelConnectedListener.java:37)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:570)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:549)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
		at org.neo4j.driver.internal.shaded.io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:263)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
		at org.neo4j.driver.internal.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
		at org.neo4j.driver.internal.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
		at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.neo4j.driver.internal.shaded.io.netty.channel.ConnectTimeoutException: connection timed out: example.com/93.184.216.34:7687
	at org.neo4j.driver.internal.shaded.io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261)
	at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at org.neo4j.driver.internal.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at org.neo4j.driver.internal.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/impl/BoltNeo4jDriverImpl.java
+++ b/neo4j-jdbc-bolt/src/main/java/org/neo4j/jdbc/bolt/impl/BoltNeo4jDriverImpl.java
@@ -65,6 +65,7 @@ public abstract class BoltNeo4jDriverImpl extends Neo4jDriver {
         Connection connection = null;
         if (acceptsURL(url)) {
             String boltUrl = url.replace(Neo4jDriver.JDBC_PREFIX, "").replaceAll("^(" + getPrefix() + ":)([^/])", "$1//$2");
+            Driver driver = null;
             try {
                 Properties info = mergeUrlAndInfo(boltUrl, props);
 
@@ -89,10 +90,17 @@ public abstract class BoltNeo4jDriverImpl extends Neo4jDriver {
                 Properties routingContext = getRoutingContext(boltUrl, info);
                 boltUrl = addRoutingPolicy(boltUrl, routingContext);
                 List<URI> routingUris = buildRoutingUris(boltUrl, routingContext);
-                Driver driver = getDriver(routingUris, config, authToken, info);
+                driver = getDriver(routingUris, config, authToken, info);
                 driver.verifyConnectivity();
                 connection = BoltNeo4jConnectionImpl.newInstance(driver, info, url);
             } catch (Exception e) {
+                if (driver != null) {
+                    try {
+                        driver.close();
+                    } catch (Exception closeException) {
+                        e.addSuppressed(closeException);
+                    }
+                }
                 throw new SQLException(e);
             }
         }

--- a/neo4j-jdbc-bolt/src/test/java/org/neo4j/jdbc/bolt/BoltNeo4jConnectionIT.java
+++ b/neo4j-jdbc-bolt/src/test/java/org/neo4j/jdbc/bolt/BoltNeo4jConnectionIT.java
@@ -20,25 +20,40 @@
 package org.neo4j.jdbc.bolt;
 
 import org.junit.After;
+import org.junit.Assert;
 import org.junit.Before;
 import org.junit.ClassRule;
 import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
+import org.neo4j.driver.AuthToken;
+import org.neo4j.driver.Config;
+import org.neo4j.driver.Driver;
+import org.neo4j.driver.GraphDatabase;
+import org.neo4j.driver.Session;
 import org.neo4j.harness.junit.rule.Neo4jRule;
 import org.neo4j.jdbc.bolt.data.StatementData;
+import org.neo4j.jdbc.bolt.impl.BoltNeo4jDriverImpl;
 import org.neo4j.jdbc.bolt.utils.JdbcConnectionTestUtils;
 
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.sql.Connection;
 import java.sql.DatabaseMetaData;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
 
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.instanceOf;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 
 /**
@@ -303,4 +318,65 @@ public class BoltNeo4jConnectionIT {
 			}
 		}
 	}
+
+    @Test
+    public void closesDriverWhenConnectivityTestFails() {
+        CustomTestBoltDriver testJdbcDriver = new CustomTestBoltDriver();
+
+        assertConnectionFails(testJdbcDriver, "jdbc:neo4j:bolt://example.com");
+        assertBoltDriverIsClosed(testJdbcDriver.getDriver());
+    }
+
+    private void assertConnectionFails(CustomTestBoltDriver testJdbcDriver, String uri) {
+        try {
+            testJdbcDriver.connect(uri, new Properties());
+            Assert.fail("Connection should fail");
+        } catch (Exception e) {
+            assertThat(e, instanceOf(SQLException.class));
+            assertThat(e.getMessage(), containsString("Unable to connect to example.com:7687"));
+        }
+    }
+
+    private void assertBoltDriverIsClosed(Driver driver) {
+        try (Session ignore = driver.session()) {
+            Assert.fail("Driver should be closed");
+        } catch (Exception e) {
+            assertThat(e, instanceOf(IllegalStateException.class));
+            assertEquals("This driver instance has already been closed", e.getMessage());
+        }
+    }
+
+    static class CustomTestBoltDriver extends BoltNeo4jDriverImpl {
+
+        private Driver driver;
+
+        protected CustomTestBoltDriver() {
+            super("bolt");
+        }
+
+        @Override
+        protected Driver getDriver(List<URI> routingUris, Config config, AuthToken authToken, Properties info) throws URISyntaxException {
+            driver = GraphDatabase.driver(routingUris.get(0), authToken, config);
+            return driver;
+        }
+
+        @Override
+        protected Properties getRoutingContext(String url, Properties properties) {
+            return new Properties();
+        }
+
+        @Override
+        protected String addRoutingPolicy(String url, Properties properties) {
+            return url;
+        }
+
+        @Override
+        protected List<URI> buildRoutingUris(String boltUrl, Properties properties) throws URISyntaxException {
+            return Collections.singletonList(new URI(boltUrl));
+        }
+
+        public Driver getDriver() {
+            return driver;
+        }
+    }
 }

